app/models/client.rb

```diff
@@ -32,11 +32,17 @@ def updated_grant_for(user, time_to_live)
     grant
   end
 
-  def get_redirect_uri(redirect_uri, query_params = nil, hash_params = nil)
+  def check_redirect_uri(redirect_uri, extra_error_msg = "")
     unless redirect_uris && redirect_uris.split(" ").include?(redirect_uri)
       # Wrong redirect URI, we should NOT redirect back to the client.
-      raise "Unauthorized redirect_uri: #{redirect_uri}. Requested query_params: #{query_params}, hash_params: #{hash_params}"
+      raise "Unauthorized redirect_uri: #{redirect_uri}. #{extra_error_msg}"
     end
+  end
+
+  def get_redirect_uri(redirect_uri, query_params = nil, hash_params = nil)
+    
+    check_redirect_uri(redirect_uri, "Requested query_params: #{query_params}, hash_params: #{hash_params}")
+
     uri = URI.parse(redirect_uri)
     if uri.fragment
```

Comment:
Shouldn't we move this check to `check_redirect_uri` too? But then `get_redirect_uri` will be empty.

Reply:
Good point. I went ahead and moved it. There is still the code to set the query and fragment params, so it makes sense to keep the two methods. Note that check is called by the validate method too.